### PR TITLE
Pertinent profiles articles/translations, fix #385

### LINKED ISSUE #385: US#40 - Consulter toutes mes publications/traductions dans mon profil
ETQ éditrice/traductrice je souhaite présenter l'ensemble de mes articles dans mon profil afin de valoriser mes contributions

- [x] doit pouvoir afficher toutes les contributions/traductions que j'ai rédigées
- [x] ne doit pas afficher les traductions que je n'ai pas réalisées
- [x] ne doit pas afficher les contributions que je n'ai pas rédigées
- [x] doit afficher séparément mes contributions et mes traductions

### PR DESCRIPTION
**Pull request by davidbgk**
And the ability to display the whole list of contributions on demand 💣

### CODE REVIEW

**Comment by davidbgk** [comment]
@cafebabel-webmaster I need you to style the page with the duality articles/translations and the footer button to display the full list.

**Comment by johangiraud** [comment]
Is is possible to keep translations and articles separated? I would like the full list button to display or full translation list or full article list @davidbgk Cause we'll have two tabs a bit like that:
180404 sketch cfbb website profil b2

**Comment by johangiraud** [comment]
Here https://github.com/cafebabel/cafebabel.com/pull/405/commits/3e673552599e5093f4b628bab9b9d9e5a514ff1c I guess there is a better way to factor fadeIn and fadeOut functions. Feel free to refactor them yourself if you like 😄 
Also for now I can't get my translations. Translation column stays empty. Is it normal?

File: cafebabel/static/js/profile-contribution-switch.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+const translationSwitcher = document.querySelector('.translation-switcher'),
+  articleSwitcher = document.querySelector('.article-switcher'),
```
Comment (davidbgk) [inline cafebabel/static/js/profile-contribution-switch.js]: Use `const` on each line.

File: cafebabel/static/js/profile-contribution-switch.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  articleSwitcher.classList.add('active')
+  /*translationList.style.opacity = 0*/
+  fadeOut(translationList)
+  fadeIn(articleList)
+})
+
+/* http://youmightnotneedjquery.com/ */
+function fadeIn(element) {
+  element.style.opacity = 0
+  var last = +new Date()
```
Comment (davidbgk) [inline cafebabel/static/js/profile-contribution-switch.js]: Use `const`.

File: cafebabel/static/js/profile-contribution-switch.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  var tick = function() {
+    element.style.opacity = +element.style.opacity + (new Date() - last) / 400
```
Comment (davidbgk) [inline cafebabel/static/js/profile-contribution-switch.js]: Explain `400`.

File: cafebabel/static/js/profile-contribution-switch.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    last = +new Date()
+    if (+element.style.opacity < 1) {
+      ;(window.requestAnimationFrame && requestAnimationFrame(tick)) ||
+        setTimeout(tick, 16)
```
Comment (davidbgk) [inline cafebabel/static/js/profile-contribution-switch.js]: Explain `16`.

**Comment by davidbgk** [comment]
@cafebabel-webmaster here is a great resource to handle tabbed and accessible interfaces https://inclusive-components.design/tabbed-interfaces/

File: cafebabel/static/js/profile-contribution-switch.js (hunk 1 above)
Comment (johangiraud) [reply cafebabel/static/js/profile-contribution-switch.js]: ok, why should I use `const` on each line?

File: cafebabel/static/js/profile-contribution-switch.js (hunk 2 above)
Comment (johangiraud) [reply cafebabel/static/js/profile-contribution-switch.js]: If I use `const` I get `Uncaught TypeError: Assignment to constant variable.` 😢

File: cafebabel/static/js/profile-contribution-switch.js (hunk 3 above)
Comment (johangiraud) [reply cafebabel/static/js/profile-contribution-switch.js]: 400 ms = fadein/fadeOut duration

File: cafebabel/static/js/profile-contribution-switch.js (hunk 4 above)
Comment (johangiraud) [reply cafebabel/static/js/profile-contribution-switch.js]: I can't explain it. So I delete it.

**Comment by johangiraud** [comment]
> @cafebabel-webmaster here is a great resource to handle tabbed and accessible interfaces https://inclusive-components.design/tabbed-interfaces/

Thanks, oh it's quite long, I'll have a look at it tomorrow

File: cafebabel/static/js/profile-contribution-switch.js (hunk 2 above)
Comment (davidbgk) [reply cafebabel/static/js/profile-contribution-switch.js]: Alright, I missed the following redeclaration, use `let` in this case.

**Comment by davidbgk** [comment]
@cafebabel-webmaster translations are present but the opacity is wrong so you cannot see the cards!

File: cafebabel/static/js/profile-contribution-switch.js (hunk 2 above)
Comment (johangiraud) [reply cafebabel/static/js/profile-contribution-switch.js]: ok

**Comment by johangiraud** [comment]
ah ok, there was a `display` issue.. sorry

File: cafebabel/templates/users/detail.html [hunk 5]
```diff
@@ -43,9 +48,20 @@ <h2>About</h2>
 </section>
 
 <section>
-  <h2>Contributions</h2>
-  {% for article in articles %}
-    {{ card(article) }}
-  {% endfor %}
+  <h2 class="contribution-switcher article-switcher active">Articles</h2><h2 class="contribution-switcher translation-switcher">Translations</h2>
```
Comment (davidbgk) [inline cafebabel/templates/users/detail.html]: Is that possible to put the Translations title on top of the `translation-list` to remain consistent HTML wise? And then adapt with CSS to keep the same look.

File: cafebabel/static/css/styles.css [hunk 6]
```diff
@@ -1622,6 +1622,31 @@ select#author {
...
+  width: 50%;
+  font-size: 2.5rem;
+  padding: 1rem;
+  margin-top: 0;
+  cursor: pointer;
+  border: 0.1rem solid #fff;
+  transition: all 0.3s;
+}
+
+.profile-page main section:nth-of-type(2) h2:nth-of-type(2) {
```
Comment (vinyll) [inline cafebabel/static/css/styles.css]: 🙃

File: cafebabel/static/css/styles.css [hunk 7]
```diff
@@ -1630,7 +1655,7 @@ select#author {
     grid-template-areas:
       'title title'
       'content sidebar';
-    padding-right: 2rem; /* don't understand why I need that. Otherwise h1 doesn't have a right margin - maybe ask Éric*/
+    padding-right: 2rem; /* don't understand why I need that. Otherwise h1 doesn't have a right margin */
```
Comment (vinyll) [inline cafebabel/static/css/styles.css]: 🙃🙂🙃

**Review by johangiraud — APPROVED** [review]

File: cafebabel/static/js/profile-contribution-switch.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
+const translationSwitcher = document.querySelector('.translation-switcher')
```
Comment (vinyll) [inline cafebabel/static/js/profile-contribution-switch.js]: I wonder if we really need this script.
The idea is to fade in and out the boxes, right?
I guess [this could do the job too](https://www.cssscript.com/creating-a-stylish-tabs-control-using-pure-css-css3/) in pure CSS, or be an inspiration.

File: cafebabel/tests/test_profile.py [hunk 9]
```diff
@@ -14,20 +14,63 @@ def test_confirm_user_creates_default_profile(app):
...
+def test_author_profile_has_list_of_published_translations_and_drafts(
+        client, user, translation, published_translation):
+    login(client, user.email, 'password')
+    response = client.get(f'/en/profile/{translation.authors[0].id}/')
+    assert response.status_code == HTTPStatus.OK
+    assert translation.detail_url in response
+    assert published_translation.detail_url in response
+
+
+def test_author_profile_has_list_of_its_translations_only(
```
Comment (vinyll) [inline cafebabel/tests/test_profile.py]: …of_his_translations…

File: cafebabel/users/views.py [hunk 10]
```diff
@@ -76,8 +77,22 @@ def edit(id):
...
-    filters = {'authors__in': [user]}
+    is_full = 'full' in request.args
+    filters = {}
     if not user.is_me():
         filters['status'] = 'published'
-    articles = Article.objects(**filters).hard_limit()
-    return render_template('users/detail.html', user=user, articles=articles)
+    articles = Article.objects(original_article=None, authors__in=[user],
+                               **filters)
+    # PERF: we exclude the `original_article` to avoid an extra query
```
Comment (vinyll) [inline cafebabel/users/views.py:86]: Brainf**ker :)

**Review by vinyll — COMMENTED** [review]
Not sure about the JS vs CSS for fading a toggle.
If it's an equivalent result as the one provided in the link, it would definitely be less bug prone!

File: cafebabel/static/js/profile-contribution-switch.js (hunk 8 above)
Comment (johangiraud) [reply cafebabel/static/js/profile-contribution-switch.js]: ok

Comment (johangiraud) [reply cafebabel/static/js/profile-contribution-switch.js]: I don't like it but I did it. Since I couldn't have it work with opacity transition, I played with display only 👎 51bd14b

File: cafebabel/templates/users/detail.html [hunk 11]
```diff
@@ -43,9 +43,27 @@ <h2>About</h2>
 </section>
 
 <section>
-  <h2>Contributions</h2>
-  {% for article in articles %}
-    {{ card(article) }}
-  {% endfor %}
+  <span>
```
Comment (davidbgk) [inline cafebabel/templates/users/detail.html]: I think that should be a `div`?

Comment (johangiraud) [reply cafebabel/templates/users/detail.html]: 👍 https://github.com/cafebabel/cafebabel.com/pull/405/commits/e826186edcf58ba21540c8a6baaad5a413194e19

**Review by vinyll — CHANGES_REQUESTED** [review]
The homepage toggle is broken when I don't have translations or articles
screen shot 2018-04-09 at 4 26 50 pm

File: cafebabel/templates/users/detail.html [hunk 12]
```diff
@@ -43,9 +43,29 @@ <h2>About</h2>
...
+  {% if translations %}
+    <input type=radio name=contribution-type-switch id=translation-list {% if not articles %}checked{% endif %}>
+    <label for=translation-list>Translations</label>
+    <div class=contributions>
+      {% for translation in translations %}
+        {{ card(translation) }}
+      {% endfor %}
+    </div>
+  {% endif %}
+  {% if articles|length >= 25 or translations|length >= 25 %}
```
Comment (davidbgk) [inline cafebabel/templates/users/detail.html]: I think we should swap this line with the one below for performances matter, this way we do not count all articles/translations when they are all displayed (which might be costly).

And the test should be `if articles|length == 25 or translations|length == 25` or is there a edge-case I forget?

Comment (johangiraud) [reply cafebabel/templates/users/detail.html]: I don't get what you mean? @davidbgk Can you swap the line yourself please?
And I don't know how to fix the test either, where should I put this line: `if articles|length == 25 or translations|length == 25`?

**Comment by johangiraud** [comment]
> Oh cool that grid would do the job. Looks explicit this way! :)

Ou! merci Vincent

**Review by vinyll — APPROVED** [review]
